### PR TITLE
Integrate manufacturer into catalogue items v2 #108

### LINKED ISSUE #108: Intergrate manufacturer into the catalogue item dialog

### LINKED ISSUE #129: Create manufacturer landing page
Create landing pages like in #75 for manufacturers

### PR DESCRIPTION
**Pull request by MatteoGuarnaccia5**
## Description

Integrated manufacturer into catalogue items, changed dialog so that users are able to select a manufacturer from a autocomplete dropdown. they can also add a new manufacturer, which will then be added to dropdown list. The ID of the manufacturer is then passed through to the backend to be added to the catalogue item. Currently the table view displays all the fields of a manufacturer (name, address ,url, telephone number) but this is subject to change based upon CLFs feedback

Note - #132 merged into this PR. So this PR will essentially close it and the issue for it

## Testing instructions

Add a set up instructions describing how the reviewer should test the code

- [ ] Review code
- [ ] Check Actions build
- [ ] Review changes to test coverage

## Agile board tracking

closes #108
closes #129

### CODE REVIEW

File: src/catalogue/items/catalogueItemsTable.component.tsx [hunk 1]
```diff
@@ -66,6 +68,8 @@ const CatalogueItemsTable = (props: CatalogueItemsTableProps) => {
 
   const { data, isLoading } = useCatalogueItems(parentInfo.id);
 
+  const { data: manufacturerList } = useManufacturers();
```
Comment (joelvdavies) [inline src/catalogue/items/catalogueItemsTable.component.tsx:71]: It would be best to only load the unique manufactures - we may end up with a lot of them. This would probably involve using a typescript `Set` to obtain ids once the catalogue items data is loaded, then making use of react-query's `useQueries` method to obtain the results for these before matching them like you have already done.

File: src/catalogue/items/catalogueItemsTable.component.tsx [hunk 2]
```diff
@@ -271,35 +275,127 @@ const CatalogueItemsTable = (props: CatalogueItemsTableProps) => {
...
+            }
           </MuiLink>
         ),
       },
+      {
+        header: 'Manufacturer Address',
+        accessorFn: (row) =>
+          `${
+            manufacturerList?.find((manufacturer) => {
+              return manufacturer.id === row.manufacturer_id;
```
Comment (joelvdavies) [inline src/catalogue/items/catalogueItemsTable.component.tsx:327]: There are quite a few lines like this missing coverage - I think its because it doesn't wait for them to be loaded.

File: src/manufacturer/manufacturerLandingPage.component.tsx [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+          onClick={() => {
+            setEditManufacturerDialogOpen(true);
+          }}
+        >
+          Edit
+        </Button>
+      </Grid>
+      {manufacturerData && (
+        <Grid item xs={12}>
+          <Box
```
Comment (joelvdavies) [inline src/manufacturer/manufacturerLandingPage.component.tsx:55]: This and the other boxes should probably be `Grid item`s, the one above would then need to be `Grid item container`, putting the `px: '192px'` in that should also avoid repeating it (like https://github.com/ral-facilities/inventory-management-system/blob/49c87c570038916d40083b49b5fa0b17c775e9d5/src/catalogue/items/catalogueItemsLandingPage.component.tsx#L115-L125)

I also think the URL/Telephone number headings should align with the address
image

File: src/catalogue/items/catalogueItemsDialog.component.tsx [hunk 4]
```diff
@@ -804,82 +778,68 @@ function CatalogueItemsDialog(props: CatalogueItemsDialogProps) {
...
+                    ) => {
+                      setSelectedManufacturer(newManufacturer ?? null);
+                      setInputValue(newManufacturer?.name ?? '');
+                      handleCatalogueDetails(
+                        'manufacturer_id',
+                        newManufacturer?.id ?? null
+                      );
+                      setManufacturerErrorMessage(undefined);
+                    }}
+                    disablePortal
```
Comment (joelvdavies) [inline src/catalogue/items/catalogueItemsDialog.component.tsx:809]: Why is this needed?

**Review by joelvdavies — COMMENTED** [review]
Some things I noticed that are mostly code related. I also noticed this behavour of the autocomplete:
autocompleteBehaviour
By typing `D` - it autocompletes so that when I type a second character `y` it ends up with a manufacturer `Digital Dynamics Incy` when I was expecting it to autocomplete to `Dynamic Designs Inc` which also exists.

File: src/catalogue/items/CatalogueItemsDetailsPanel.component.tsx [hunk 5]
```diff
@@ -268,60 +274,79 @@ function CatalogueItemsDetailsPanel(props: CatalogueItemsDetailsPanelProps) {
...
+                      <MuiLink
+                        underline="hover"
+                        target="_blank"
+                        href={manufacturerData?.url}
+                      >
+                        {manufacturerData?.url ?? 'None'}
+                      </MuiLink>
+                    }
+                  />
+                </ListItem>
```
Comment (joelvdavies) [inline src/catalogue/items/CatalogueItemsDetailsPanel.component.tsx:307]: Could you also apply the ListItem change to Typographies to these and the others in the file (assuming they have no parent list) - they are all the same as you have already done.

File: src/catalogue/items/catalogueItemsDialog.component.test.tsx [hunk 6]
```diff
@@ -155,6 +138,18 @@ describe('Catalogue Items Dialog', () => {
...
+  const modifyManufacturer = async (type: string) => {
+    const manufacturerPopup = screen.getAllByRole('combobox')[0];
+    await user.type(
+      manufacturerPopup,
+      type === 'create'
+        ? 'M{arrowdown}{enter}'
+        : 'Man{arrowdown}{arrowdown}{enter}'
+    );
+    const saveButton = screen.getByRole('button', { name: 'Save' });
+    await user.click(saveButton);
```
Comment (joelvdavies) [inline src/catalogue/items/catalogueItemsDialog.component.test.tsx:150]:
```suggestion
if (values.manufactuer !== undefined) {
    const manufacturerPopup = screen.getAllByRole('combobox')[0];
    await user.type(
      manufacturerPopup,
      manufacturer
    );
    const saveButton = screen.getByRole('button', { name: 'Save' });
    await user.click(saveButton);
 }
```
Then this can be put into the `modifyValues` and you can use it instead of making a separate call to `modifyManufacturer`

File: src/catalogue/items/catalogueItemsLandingPage.component.tsx [hunk 7]
```diff
@@ -422,65 +427,69 @@ function CatalogueItemsLandingPage() {
...
-                          flexDirection: 'row',
-                          alignItems: 'center',
-                        }}
-                      >
-                        <ListItemText
-                          primary={'Manufacturer Address'}
-                          secondary={catalogueItemIdData.manufacturer.address}
-                        />
-                      </ListItem>
+              {manufacturer && (
```
Comment (joelvdavies) [inline src/catalogue/items/catalogueItemsLandingPage.component.tsx:430]: Could you also modify the other ListItem's in this file above (when they don't have a parent List) please

File: src/manufacturer/manufacturerLandingPage.component.tsx [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+                  alignItems: 'center',
+                  cursor: 'pointer',
+                }}
+              >
+                <Typography variant="h6">Address</Typography>
+              </Grid>
+
+              <Grid container spacing={1}>
+                <Grid item xs={12} sm={6} md={4}>
+                  <ListItem
```
Comment (joelvdavies) [inline src/manufacturer/manufacturerLandingPage.component.tsx:133]: Some more ListItems here.

File: src/manufacturer/manufacturerLandingPage.component.tsx [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+            </Grid>
+            <Grid container spacing={1} sx={{ px: '192px' }}>
+              <Grid
+                item
+                xs={12}
+                sx={{
+                  display: 'flex',
+                  flexDirection: 'row',
+                  alignItems: 'center',
+                  cursor: 'pointer',
```
Comment (joelvdavies) [inline src/manufacturer/manufacturerLandingPage.component.tsx:69]:
```suggestion
```
There are several of these below, they make it seem like something is clickable when it is not.

File: src/catalogue/items/catalogueItemsLandingPage.component.tsx [hunk 10]
```diff
@@ -422,65 +427,69 @@ function CatalogueItemsLandingPage() {
...
+                        </Typography>
+                      </Grid>
+                      <Grid item xs={12}>
+                        <Typography color="text.primary">
+                          Post/Zip code
+                        </Typography>
+                        <Typography color="text.secondary">
+                          {manufacturer?.address.postcode}
+                        </Typography>
+                      </Grid>
```
Comment (joelvdavies) [inline src/catalogue/items/catalogueItemsLandingPage.component.tsx:488]: I think these probably need styles to match the formatting above:
image

**Review by joelvdavies — CHANGES_REQUESTED** [review]
A couple more things for consistency. Works well for me though!

File: src/manufacturer/manufacturerDialog.component.tsx [hunk 11]
```diff
@@ -63,7 +63,7 @@ function ManufacturerDialog(props: ManufacturerDialogProps) {
 
   React.useEffect(() => {
     if (selectedManufacturer) setManufacturerDetails(selectedManufacturer);
-  }, [selectedManufacturer]);
+  }, [selectedManufacturer, open]);
```
[the remark this answers is not in the document]
Comment (joshuadkitenge) [reply src/manufacturer/manufacturerDialog.component.tsx]: This is needed for the edit dialog on the landing page. or it cause the bug below 

man ufacturer

<br><br>

Also the dialog should be using handleClose not onClose 

```
 <Dialog open={open} onClose={onClose} maxWidth="lg" fullWidth>
      <DialogTitle>{`${
        !selectedManufacturer ? 'Add' : 'Edit'
      } Manufacturer`}</DialogTitle>
...
```

File: src/catalogue/items/catalogueItemsDialog.component.tsx [hunk 12]
```diff
@@ -804,82 +780,70 @@ function CatalogueItemsDialog(props: CatalogueItemsDialogProps) {
...
+                        }
+                      />
+                    )}
+                  />
+                </Grid>
+                <Grid item xs={1}>
+                  <IconButton
+                    sx={{ mx: '4px', my: '2px' }}
+                    onClick={() => setAddManufacturerDialogOpen(true)}
+                    disabled={selectedManufacturer !== null}
```
Comment (joshuadkitenge) [inline src/catalogue/items/catalogueItemsDialog.component.tsx:836]: Why is this button being disabled?
It is not consistent as it disables the button when you selected a manufacturer when your adding a catalogue item but with save as and edit the button is enabled. look at gif below 

  
man ufacturer add in cat

File: src/catalogue/items/catalogueItemsDialog.component.test.tsx [hunk 13]
```diff
@@ -16,7 +16,7 @@ import userEvent from '@testing-library/user-event';
 import CatalogueItemsDialog, {
   CatalogueItemsDialogProps,
 } from './catalogueItemsDialog.component';
-
+jest.setTimeout(10000); //multiple long running tests
```
Comment (joshuadkitenge) [inline src/catalogue/items/catalogueItemsDialog.component.test.tsx:19]: Does the test still time out when ran on the CI or can this be removed ??

**Review by joshuadkitenge — CHANGES_REQUESTED** [review]

File: src/catalogue/items/catalogueItemsDialog.component.test.tsx (hunk 6 above)
Comment (MatteoGuarnaccia5) [reply src/catalogue/items/catalogueItemsDialog.component.test.tsx:150]: I tried doing this initially, but it just does not seem to work. The tests fail for a variety of reasons. Maybe it has to do with the fact I'm using user and not fireEvent

**Comment by MatteoGuarnaccia5** [comment]
I tried to simplify the logic for selectedManufacturer in the Dialog. The solution I found doesn't seem to work when you then navigate to the landing page and try to edit an item from there. I can keep trying if you guys wish but it is as compact as I can make it

**Comment by joelvdavies** [comment]
> I tried to simplify the logic for selectedManufacturer in the Dialog. The solution I found doesn't seem to work when you then navigate to the landing page and try to edit an item from there. I can keep trying if you guys wish but it is as compact as I can make it

If this is about the comment I made in catalogueItemsDialog, replacing
`
  const { data: selectedCatalogueItemManufacturer } = useManufacturer(
    selectedCatalogueItem?.manufacturer_id
  );
`
with
`
  const selectedCatalogueItemManufacturer =
    manufacturerList?.find(
      (manufactuerer) =>
        manufactuerer.id === selectedCatalogueItem?.manufacturer_id
    ) || null;
`
appears to work. But don't worry about it if not.

**Review by joelvdavies — CHANGES_REQUESTED** [review]
Last two things from me
The address on the manufacturers landing page is well spaced out
image

But on the catalogue items page it is not
image

There are also still a few ListItems in `catalogueItemsLandingPage.component.tsx`.

**Review by joshuadkitenge — CHANGES_REQUESTED** [review]
The catalogue Item landing page styling needs to be same as what it was before but using typographys 

Before:
image


After: 
image


This styling will also need to carry through into the manufacturer landing page so that it looks similar to the catalogue items landing page

**Review by joshuadkitenge — CHANGES_REQUESTED** [review]
Also, can you change how the manufacturer address is displayed so it matches what is in the table. below:
image
This format in both of the landing pages.

As currently the address is not in the most readable format
image

**Review by joshuadkitenge — DISMISSED** [review]

**Comment by MatteoGuarnaccia5** [comment]
Needs to be merged in first priority after christmas

**Review by joshuadkitenge — APPROVED** [review]
LGTM

**Review by joelvdavies — APPROVED** [review]
